aws/hll/aws-s3-transfer-manager/src/io/path_body.rs

```diff
@@ -0,0 +1,210 @@
+    ///
+    /// When used in conjunction with [`offset`](PathBodyBuilder::offset), allows for reading a single "chunk" of a file.
+    ///
+    /// <div class="warning">
+    /// Setting the length manually will trigger no validation related to any offset provided or the actual size of
+    /// the file. This is an advanced setting mainly used to avoid an additional syscall if you know the
+    /// size of the file already.
+    /// </div>
+    pub fn length(mut self, length: u64) -> Self {
+        self.length = Some(length);
+        self
+    }
+
+    /// Returns a [`InputStream`] from this builder.
+    pub fn build(self) -> Result<InputStream, Error> {
+        let path = self.path.expect("path set");
+        let offset = self.offset.unwrap_or_default();
+
+        let length = match self.length {
+            None => {
+                // TODO(aws-sdk-rust#1159, design) - evaluate if we want build() to be async and to use tokio for stat() call (bytestream FsBuilder::build() is async)
+                let metadata = fs::metadata(path.clone())?;
+                let file_size = metadata.len();
+
+                if offset >= file_size {
+                    return Err(ErrorKind::OffsetGreaterThanFileSize.into());
+                }
+
+                file_size - offset
+            }
+            Some(explicit) => explicit,
+        };
+
+        let body = PathBody {
+            path,
+            length,
+            offset,
+        };
+
+        let stream = InputStream {
+            inner: RawInputStream::Fs(body),
+        };
+
+        Ok(stream)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Write;
+    use tempfile::NamedTempFile;
+
+    use crate::io::{path_body::PathBodyBuilder, InputStream};
+
+    use super::PathBody;
+
+    fn path_body(stream: &InputStream) -> &PathBody {
+        match &stream.inner {
+            crate::io::stream::RawInputStream::Buf(_) => panic!("unexpected inner body"),
+            crate::io::stream::RawInputStream::Fs(path_body) => path_body,
+        }
+    }
+
+    #[test]
+    fn test_from_path() {
+        let mut tmp = NamedTempFile::new().unwrap();
+        let content = "hello path body";
+        tmp.write_all(content.as_bytes()).unwrap();
+
+        let stream = PathBodyBuilder::new().path(tmp.path()).build().unwrap();
+        let body = path_body(&stream);
+        assert_eq!(0, body.offset);
+        assert_eq!(content.as_bytes().len() as u64, body.length);
+    }
+
+    #[test]
+    fn test_explicit_content_length() {
+        let mut tmp = NamedTempFile::new().unwrap();
+
+        let stream = PathBodyBuilder::new()
+            .path(tmp.path())
+            .length(64)
+            .build()
+            .unwrap();
+
+        let body = path_body(&stream);
+        assert_eq!(0, body.offset);
+        // we don't validate this
+        assert_eq!(64, body.length);
```

Comment:
Could you also note "why not?", I'm curious

Reply:
When we get to implementing an actual transfer manager and are traversing a directory I expect we'll already have this metadata available and we can avoid any additional syscall. We could make this `pub(crate)` potentially.